File: python/tests/BookListingFormatterTest.py
import unittest
from tdd.book_listing_formatter import BookListingFormatter
from tdd.book import Book

class BookListingFormatterTest(unittest.TestCase):

    def test_should_display_title_followed_by_author_when_no_ratings_or_prizes(self):
        book = Book("The Hobbit", "J. R. R. Tolkien")
        listing = BookListingFormatter.create_listing_text(book, book.author, None, None, None)
        self.assertEqual(listing, "The Hobbit by J. R. R. Tolkien")

    def test_should_display_title_followed_by_author_and_rating_when_no_prizes_and_rating_is_1(self):
        book = Book("The Hobbit", "J. R. R. Tolkien")
        listing = BookListingFormatter.create_listing_text(book, book.author, 1, None, None)
        self.assertEqual(listing, "The Hobbit by J. R. R. Tolkien: 1 star")

    def test_should_display_title_followed_by_author_and_rating_when_no_prizes_and_rating_is_5(self):
        book = Book("The Hobbit", "J. R. R. Tolkien")
        listing = BookListingFormatter.create_listing_text(book, book.author, 5, None, None)
        self.assertEqual(listing, "The Hobbit by J. R. R. Tolkien: 5 stars")

    def test_should_prefix_title_with_prize_winning_when_book_supplied(self):
        book = Book("The Hobbit", "J. R. R. Tolkien")
        listing = BookListingFormatter.create_listing_text(book, book.author, 5, "Fancy Prize", None)
        self.assertEqual(listing, "Fancy Prize winner - The Hobbit by J. R. R. Tolkien: 5 stars")

    def test_should_append_winner_of_author_prize_when_supplied(self):
        book = Book("The Hobbit", "J. R. R. Tolkien")
        listing = BookListingFormatter.create_listing_text(book, book.author, 5, "Fancy Prize", "Awesome Award")
        self.assertEqual(listing, "Fancy Prize winner - The Hobbit by J. R. R. Tolkien (Winner of Awesome Award): 5 stars")

if __name__ == "__main__":
    unittest.main()

File: typescript/src/BookListingFormatter.test.ts
import { Book } from './Book';

import { BookListingFormatter } from './BookListingFormatter';

describe('BookListingFormatter', () => {
  it('should display title followed by author when no ratings or prizes', () => {
    const book = new Book('The Hobbit', 'J. R. R. Tolkien');
    const listing = BookListingFormatter.createListingText(book, book.getAuthor(), null, null, null);
    expect(listing).toBe('The Hobbit by J. R. R. Tolkien');
  });

  it('should display title followed by author and rating when no prizes and rating is 1', () => {
    const book = new Book('The Hobbit', 'J. R. R. Tolkien');
    const listing = BookListingFormatter.createListingText(book, book.getAuthor(), 1, null, null);
    expect(listing).toBe('The Hobbit by J. R. R. Tolkien: 1 star');
  });

  it('should display title followed by author and rating when no prizes and rating is 5', () => {
    const book = new Book('The Hobbit', 'J. R. R. Tolkien');
    const listing = BookListingFormatter.createListingText(book, book.getAuthor(), 5, null, null);
    expect(listing).toBe('The Hobbit by J. R. R. Tolkien: 5 stars');
  });

  it('should prefix title with prize winning when book supplied', () => {
    const book = new Book('The Hobbit', 'J. R. R. Tolkien');
    const listing = BookListingFormatter.createListingText(book, book.getAuthor(), 5, 'Fancy Prize', null);
    expect(listing).toBe('Fancy Prize winner - The Hobbit by J. R. R. Tolkien: 5 stars');
  });

  it('should append winner of author prize when supplied', () => {
    const book = new Book('The Hobbit', 'J. R. R. Tolkien');
    const listing = BookListingFormatter.createListingText(book, book.getAuthor(), 5, 'Fancy Prize', 'Awesome Award');
    expect(listing).toBe('Fancy Prize winner - The Hobbit by J. R. R. Tolkien (Winner of Awesome Award): 5 stars');
  });
});
